File: server/routes/books.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const path = require('path');
const fs = require('fs').promises;
const Book = require('../models/Book');
const RawText = require('../models/RawText');
const Summary = require('../models/Summary');
const { extractTextFromFile } = require('../utils/fileExtractor');
const { authenticateToken } = require('./auth');

// Configure multer for file uploads
const storage = multer.diskStorage({
  destination: async (req, file, cb) => {
    const uploadDir = path.join(__dirname, '../uploads');
    try {
      await fs.mkdir(uploadDir, { recursive: true });
      cb(null, uploadDir);
    } catch (error) {
      cb(error, null);
    }
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1E9);
    cb(null, uniqueSuffix + '-' + file.originalname);
  }
});

const upload = multer({
  storage: storage,
  limits: {
    fileSize: 50 * 1024 * 1024 // 50MB limit
  },
  fileFilter: (req, file, cb) => {
    const allowedExtensions = ['.pdf', '.docx', '.txt'];
    const ext = path.extname(file.originalname).toLowerCase();
    if (allowedExtensions.includes(ext)) {
      cb(null, true);
    } else {
      cb(new Error('Only PDF, DOCX, and TXT files are allowed'));
    }
  }
});

// POST /api/books/upload - Upload book file and extract text
router.post('/upload', authenticateToken, upload.single('file'), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }

    const { title, author, genre, publication_year, pages } = req.body;

    // Validation
    if (!title || !author) {
      // Clean up uploaded file
      await fs.unlink(req.file.path);
      return res.status(400).json({ 
        error: 'Missing required fields',
        required: ['title', 'author']
      });
    }

    console.log(`📄 Processing uploaded file: ${req.file.originalname}`);
    
    // Extract text from file
    const fileExtension = path.extname(req.file.originalname);
    const extractedText = await extractTextFromFile(req.file.path, fileExtension);
    
    // Create book record
    const book = new Book({
      title,
      author,
      uploaded_by: req.user._id,
      upload_date: new Date(),
      genre: genre || undefined,
      publication_year: publication_year ? parseInt(publication_year) : undefined,
      pages: pages ? parseInt(pages) : undefined,
      original_filename: req.file.originalname,
      file_size: req.file.size,
      file_type: fileExtension.substring(1), // Remove the dot
      is_processed: true
    });

    await book.save();

    // Create raw text record
    const rawText = new RawText({
      book_id: book._id,
      full_text: extractedText,
      extraction_method: fileExtension.substring(1)
    });

    await rawText.save();

    // Clean up the uploaded file after extraction
    try {
      await fs.unlink(req.file.path);
      console.log(`🗑️ Cleaned up temporary file: ${req.file.filename}`);
    } catch (cleanupError) {
      console.warn('Failed to delete temporary file:', cleanupError);
    }
    
    res.status(201).json({
      message: 'Book uploaded and processed successfully',
      book: {
        id: book._id,
        title: book.title,
        author: book.author,
        upload_date: book.upload_date,
        file_size: book.file_size,
        word_count: rawText.word_count
      }
    });
    
  } catch (error) {
    // Clean up file on error
    if (req.file && req.file.path) {
      try {
        await fs.unlink(req.file.path);
      } catch (cleanupError) {
        console.warn('Failed to delete temporary file on error:', cleanupError);
      }
    }
    
    console.error('Book upload error:', error);
    res.status(500).json({ 
      error: 'Book upload failed', 
      message: error.message 
    });
  }
});

// GET /api/books - Get all books
router.get('/', authenticateToken, async (req, res) => {
  try {
    const { search, sort = '-upload_date', limit = 50, page = 1 } = req.query;
    
    let query = {};
    
    // Search functionality
    if (search) {
      query.$text = { $search: search };
    }

    const skip = (parseInt(page) - 1) * parseInt(limit);

    const books = await Book.find(query)
      .populate('uploaded_by', 'name email')
      .sort(sort)
      .limit(parseInt(limit))
      .skip(skip)
      .select('-__v');

    const total = await Book.countDocuments(query);

    res.json({
      books,
      pagination: {
        total,
        page: parseInt(page),
        limit: parseInt(limit),
        pages: Math.ceil(total / parseInt(limit))
      }
    });
  } catch (error) {
    console.error('Fetch books error:', error);
    res.status(500).json({ 
      error: 'Failed to fetch books', 
      message: error.message 
    });
  }
});

// GET /api/books/my-books - Get books uploaded by current user
router.get('/my-books', authenticateToken, async (req, res) => {
  try {
    const books = await Book.find({ uploaded_by: req.user._id })
      .sort({ upload_date: -1 })
      .select('-__v');

    res.json({
      count: books.length,
      books
    });
  } catch (error) {
    res.status(500).json({ 
      error: 'Failed to fetch your books', 
      message: error.message 
    });
  }
});

// GET /api/books/:id - Get single book with details
router.get('/:id', authenticateToken, async (req, res) => {
  try {
    const book = await Book.findById(req.params.id)
      .populate('uploaded_by', 'name email')
      .select('-__v');

    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }

    // Get raw text info (without full text)
    const rawText = await RawText.findOne({ book_id: book._id })
      .select('word_count character_count extraction_method extraction_date');

    // Get summaries count
    const summariesCount = await Summary.countDocuments({ book_id: book._id });

    res.json({
      book,
      textInfo: rawText,
      summariesCount
    });
  } catch (error) {
    res.status(500).json({ 
      error: 'Failed to fetch book', 
      message: error.message 
    });
  }
});

// GET /api/books/:id/text - Get full text of a book
router.get('/:id/text', authenticateToken, async (req, res) => {
  try {
    const book = await Book.findById(req.params.id);
    
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }

    const rawText = await RawText.findOne({ book_id: book._id });
    
    if (!rawText) {
      return res.status(404).json({ error: 'Text not found for this book' });
    }

    res.json({
      book_id: book._id,
      title: book.title,
      full_text: rawText.full_text,
      word_count: rawText.word_count,
      character_count: rawText.character_count
    });
  } catch (error) {
    res.status(500).json({ 
      error: 'Failed to fetch book text', 
      message: error.message 
    });
  }
});

// PUT /api/books/:id - Update book metadata
router.put('/:id', authenticateToken, async (req, res) => {
  try {
    const book = await Book.findById(req.params.id);
    
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }

    // Check ownership or admin
    if (book.uploaded_by.toString() !== req.user._id.toString() && req.user.role !== 'admin') {
      return res.status(403).json({ error: 'You do not have permission to update this book' });
    }

    const { title, author, genre, publication_year, pages } = req.body;
    const updates = {};

    if (title !== undefined) updates.title = title;
    if (author !== undefined) updates.author = author;
    if (genre !== undefined) updates.genre = genre;
    if (publication_year !== undefined) updates.publication_year = publication_year;
    if (pages !== undefined) updates.pages = pages;

    const updatedBook = await Book.findByIdAndUpdate(
      req.params.id,
      updates,
      { new: true, runValidators: true }
    ).populate('uploaded_by', 'name email');

    res.json({
      message: 'Book updated successfully',
      book: updatedBook
    });
  } catch (error) {
    res.status(400).json({ 
      error: 'Failed to update book', 
      message: error.message 
    });
  }
});

// DELETE /api/books/:id - Delete book and associated data
router.delete('/:id', authenticateToken, async (req, res) => {
  try {
    const book = await Book.findById(req.params.id);
    
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }

    // Check ownership or admin
    if (book.uploaded_by.toString() !== req.user._id.toString() && req.user.role !== 'admin') {
      return res.status(403).json({ error: 'You do not have permission to delete this book' });
    }

    // Delete associated raw text
    await RawText.deleteOne({ book_id: book._id });

    // Delete associated summaries
    await Summary.deleteMany({ book_id: book._id });

    // Delete the book
    await Book.findByIdAndDelete(req.params.id);

    res.json({ 
      message: 'Book and all associated data deleted successfully',
      bookId: req.params.id
    });
  } catch (error) {
    res.status(500).json({ 
      error: 'Failed to delete book', 
      message: error.message 
    });
  }
});

// GET /api/books/:id/summaries - Get all summaries for a book
router.get('/:id/summaries', authenticateToken, async (req, res) => {
  try {
    const book = await Book.findById(req.params.id);
    
    if (!book) {
      return res.status(404).json({ error: 'Book not found' });
    }

    const summaries = await Summary.find({ book_id: book._id })
      .populate('generated_by', 'name email')
      .sort({ generation_date: -1 })
      .select('-__v');

    res.json({
      book_id: book._id,
      title: book.title,
      summaries_count: summaries.length,
      summaries
    });
  } catch (error) {
    res.status(500).json({ 
      error: 'Failed to fetch summaries', 
      message: error.message 
    });
  }
});

module.exports = router;

File: server/routes/books.py
from fastapi import APIRouter, HTTPException, Depends, UploadFile, File, Form, status
from motor.motor_asyncio import AsyncIOMotorDatabase
from datetime import datetime
from bson import ObjectId
from pathlib import Path
import os
from typing import Optional

from models.book import BookResponse
from models.raw_text import RawText
from utils.auth import get_current_user
from utils.file_extractor import extract_text_from_file

router = APIRouter()

# Configure upload directory
UPLOAD_DIR = Path(__file__).parent.parent / "uploads"
UPLOAD_DIR.mkdir(exist_ok=True)

def get_database():
    from main import database
    return database

@router.post("/upload", response_model=BookResponse, status_code=status.HTTP_201_CREATED)
async def upload_book(
    file: UploadFile = File(...),
    title: str = Form(...),
    author: str = Form(...),
    genre: Optional[str] = Form(None),
    publication_year: Optional[int] = Form(None),
    pages: Optional[int] = Form(None),
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Upload book file and extract text"""
    file_path = None
    try:
        # Validate file extension
        file_extension = Path(file.filename).suffix.lower()
        allowed_extensions = ['.pdf', '.docx', '.txt']
        if file_extension not in allowed_extensions:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Only {', '.join(allowed_extensions)} files are allowed"
            )
        
        # Validate required fields
        if not title or not author:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Title and author are required"
            )
        
        # Save uploaded file
        unique_filename = f"{int(datetime.utcnow().timestamp() * 1000)}-{file.filename}"
        file_path = UPLOAD_DIR / unique_filename
        
        with open(file_path, "wb") as f:
            content = await file.read()
            f.write(content)
        
        print(f"📄 Processing uploaded file: {file.filename}")
        
        # Extract text from file
        extracted_text = await extract_text_from_file(str(file_path), file_extension)
        
        # Create book record
        book_dict = {
            "title": title,
            "author": author,
            "uploaded_by": current_user.get("userId"),
            "upload_date": datetime.utcnow(),
            "genre": genre,
            "publication_year": publication_year,
            "pages": pages,
            "original_filename": file.filename,
            "file_size": len(content),
            "file_type": file_extension.replace('.', ''),
            "is_processed": True,
            "createdAt": datetime.utcnow(),
            "updatedAt": datetime.utcnow()
        }
        
        result = await db.books.insert_one(book_dict)
        book_id = str(result.inserted_id)
        
        # Calculate word count
        word_count = len(extracted_text.split())
        
        # Create raw text record
        raw_text_dict = {
            "book_id": book_id,
            "full_text": extracted_text,
            "word_count": word_count,
            "extraction_method": file_extension.replace('.', ''),
            "extraction_date": datetime.utcnow(),
            "createdAt": datetime.utcnow(),
            "updatedAt": datetime.utcnow()
        }
        
        await db.raw_texts.insert_one(raw_text_dict)
        
        # Clean up the uploaded file
        try:
            os.unlink(file_path)
            print(f"🗑️ Cleaned up temporary file: {unique_filename}")
        except Exception as cleanup_error:
            print(f"Warning: Failed to delete temporary file: {cleanup_error}")
        
        return BookResponse(
            id=book_id,
            title=title,
            author=author,
            upload_date=book_dict["upload_date"],
            file_size=len(content),
            word_count=word_count
        )
        
    except HTTPException:
        raise
    except Exception as e:
        # Clean up file on error
        if file_path and os.path.exists(file_path):
            try:
                os.unlink(file_path)
            except:
                pass
        
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Book upload failed: {str(e)}"
        )

@router.get("/")
async def get_all_books(
    search: Optional[str] = None,
    sort: str = "-upload_date",
    limit: int = 50,
    page: int = 1,
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Get all books"""
    try:
        query = {}
        
        if search:
            query["$or"] = [
                {"title": {"$regex": search, "$options": "i"}},
                {"author": {"$regex": search, "$options": "i"}}
            ]
        
        # Parse sort parameter
        sort_field = sort.lstrip('-')
        sort_direction = -1 if sort.startswith('-') else 1
        
        skip = (page - 1) * limit
        
        books_cursor = db.books.find(query).sort(sort_field, sort_direction).skip(skip).limit(limit)
        books = await books_cursor.to_list(length=limit)
        
        # Convert ObjectId to string
        for book in books:
            book["id"] = str(book["_id"])
            book["_id"] = str(book["_id"])
            if "uploaded_by" in book:
                book["uploaded_by"] = str(book["uploaded_by"])
        
        total = await db.books.count_documents(query)
        
        return {
            "books": books,
            "pagination": {
                "total": total,
                "page": page,
                "limit": limit,
                "pages": (total + limit - 1) // limit
            }
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch books: {str(e)}"
        )

@router.get("/my-books")
async def get_my_books(
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Get books uploaded by current user"""
    try:
        books_cursor = db.books.find({"uploaded_by": current_user.get("userId")}).sort("upload_date", -1)
        books = await books_cursor.to_list(length=1000)
        
        for book in books:
            book["id"] = str(book["_id"])
            book["_id"] = str(book["_id"])
            if "uploaded_by" in book:
                book["uploaded_by"] = str(book["uploaded_by"])
        
        return {
            "count": len(books),
            "books": books
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch your books: {str(e)}"
        )

@router.get("/{book_id}")
async def get_book_by_id(
    book_id: str,
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Get single book with details"""
    try:
        if not ObjectId.is_valid(book_id):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid book ID"
            )
        
        book = await db.books.find_one({"_id": ObjectId(book_id)})
        if not book:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Book not found"
            )
        
        book["id"] = str(book["_id"])
        book["_id"] = str(book["_id"])
        if "uploaded_by" in book:
            book["uploaded_by"] = str(book["uploaded_by"])
        
        return book
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch book: {str(e)}"
        )

@router.get("/{book_id}/text")
async def get_book_text(
    book_id: str,
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Get raw text for a book"""
    try:
        if not ObjectId.is_valid(book_id):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid book ID"
            )
        
        raw_text = await db.raw_texts.find_one({"book_id": book_id})
        if not raw_text:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Book text not found"
            )
        
        raw_text["id"] = str(raw_text["_id"])
        raw_text["_id"] = str(raw_text["_id"])
        
        return raw_text
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch book text: {str(e)}"
        )

@router.delete("/{book_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_book(
    book_id: str,
    current_user: dict = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """Delete a book"""
    try:
        if not ObjectId.is_valid(book_id):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid book ID"
            )
        
        book = await db.books.find_one({"_id": ObjectId(book_id)})
        if not book:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Book not found"
            )
        
        # Check if user owns the book or is admin
        if str(book["uploaded_by"]) != current_user.get("userId") and current_user.get("role") != "admin":
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You don't have permission to delete this book"
            )
        
        # Delete book, raw text, and summaries
        await db.books.delete_one({"_id": ObjectId(book_id)})
        await db.raw_texts.delete_many({"book_id": book_id})
        await db.summaries.delete_many({"book_id": book_id})
        
        return None
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete book: {str(e)}"
        )
